helix-term/src/commands/typed.rs

```diff
@@ -2483,6 +2483,101 @@ fn move_buffer(
     Ok(())
 }
 
+fn trim_whitespace_impl(doc: &Document, selection: &Selection) -> Transaction {
+    /// Maps reverse index to non-reverse index
+    fn map_reverse(len: usize, rev: usize) -> usize {
+        // This function won't be correct if len is 0
+        debug_assert_ne!(len, 0);
+        (len - 1) - rev
+    }
+
+    /// Find the last non-whitespace char of a line
+    fn find_trailing(l: RopeSlice) -> Option<usize> {
+        // Handle empty docs
+        if l.len_chars() == 0 {
+            return None;
+        }
+
+        // Returns the left-wise beginning of the trailing whitespace
+        // It is +1 the index of that char so that char is not deleted
+        l.chars_at(l.len_chars())
+            .reversed()
+            .position(|ch| !ch.is_whitespace())
+            .map(|n| l.len_chars() - n)
+            .or(Some(0))
+    }
+
+    let mut deletions: Vec<helix_core::Deletion> = Vec::new();
+    let mut delete = |start, end| {
+        // Don't push empty changes
+        if start != end {
+            deletions.push((start, end));
+        }
+    };
+
+    // Assume ranges are in order and not overlapping
+    for range in selection.ranges().iter().rev() {
```

Comment:
I wonder if instead of operating on each selection's range we should use `Selection::line_ranges` like we do for `:reset-diff-change` now. I don't think it ever makes sense to not operate on full lines.

Reply:
I might have misunderstood the code or this review, but here are the pros & cons that come to mind right now:
- Trimming selections is much more versatile, but comes at the cost of hitting `x` before the cmd (if we want to trim the full line)
- Full-line-trim seems more common (in practice), so it should have a shorter key-sequence. But (mathematically) it's just 1 special case among an `usize::MAX` ("pseudo-Infinity") of potential selection-trims. Therefore, full-line-trim is rare (in theory)

Although, it seems fallacious to believe that FLT is common, simply because practically-all editors only support trimming all lines in a buffer, thereby forcing users to get used to this limitation

Reply:
For a bit of a contrived example, say we have a file like:

```
····[········ffffffffffff···············
············ffffffffffff···············
············ffffffffffff·······]········
```

(with "·" being a space character and the selection being represented by the brackets)

I think `:trim` here should operate on all lines covered by the selection, resulting in

```
············ffffffffffff
············ffffffffffff
············ffffffffffff
```

where I think currently this PR would end up with

```
············ffffffffffff
············ffffffffffff
············ffffffffffff·······
```

I'm not sure it's ever useful to use the end of the selection range as the end of the area you're trimming

Reply:
I agree. Now I see why it doesn't seem useful.

> ```
> ············ffffffffffff
> ············ffffffffffff
> ············ffffffffffff·······
> ```

In my mind, I thought it would be:
```
············ffffffffffff············
············ffffffffffff············
············ffffffffffff·······
```
Which makes less sense, I guess?

Then I thought "what if there were non-space code-points somewhere between the consecutive spaces?", then I answered my own question:

In practice, typically, users want 0, 1, or (rare) 2 spaces between non-space chars (let's ignore indentation, for simplicity). So, why would users need to specify how many spaces they want to delete? It'd be [much simpler](https://en.wikipedia.org/wiki/Zero_one_infinity_rule) to:
- delete all spaces from all lines
- delete all spaces from all lines, except 1 space from the last line
- preserve some lines as-is